contracts/BatchExchangeViewer.sol

```diff
@@ -0,0 +1,120 @@
+     *  @return encoded bytes representing orders and page information for next page
+     */
+    function getFinalizedOrderBookPaginated(address previousPageUser, uint16 previousPageUserOffset, uint16 pageSize)
+        public
+        view
+        returns (bytes memory elements, address nextPageUser, uint16 nextPageUserOffset)
+    {
+        uint32 batch = batchExchange.getCurrentBatchId();
+        return getEncodedOrdersPaginated(batch - 1, batch - 1, previousPageUser, previousPageUserOffset, pageSize);
+    }
+
+    /** @dev Queries a page in the list of all orders
+     *  @param maxValidFrom all returned orders will have a validFrom <= this value (they were placed at or before that batch)
+     *  @param minValidUntil all returned orders will have a validUntil >= this value (validity ends at or after that batch)
+     *  @param previousPageUser address taken from nextPageUser return value from last page (address(0) for first page)
+     *  @param previousPageUserOffset offset taken nextPageUserOffset return value from last page (0 for first page)
+     *  @param pageSize count of elements to be returned per page (same value is used for subqueries on the exchange)
+     *  @return encoded bytes representing orders and page information for next page
+     */
+    function getEncodedOrdersPaginated(
+        uint32 maxValidFrom,
+        uint32 minValidUntil,
+        address previousPageUser,
+        uint16 previousPageUserOffset,
+        uint16 pageSize
+    ) public view returns (bytes memory elements, address nextPageUser, uint16 nextPageUserOffset) {
+        nextPageUser = previousPageUser;
+        nextPageUserOffset = previousPageUserOffset;
+        bool hasNextPage = true;
+        while (hasNextPage) {
+            bytes memory unfiltered = batchExchange.getEncodedUsersPaginated(nextPageUser, nextPageUserOffset, pageSize);
+            hasNextPage = unfiltered.length / AUCTION_ELEMENT_WIDTH == pageSize;
+            for (uint16 index = 0; index < unfiltered.length / AUCTION_ELEMENT_WIDTH; index++) {
+                bytes memory element = unfiltered.slice(index * AUCTION_ELEMENT_WIDTH, AUCTION_ELEMENT_WIDTH);
+                if (maxValidFrom >= getValidFrom(element) && minValidUntil <= getValidUntil(element)) {
+                    elements = elements.concat(element);
+                }
+                // Update pagination info
+                address user = getUser(element);
+                if (user == nextPageUser) {
+                    nextPageUserOffset += 1;
+                } else {
+                    nextPageUserOffset = 1;
+                    nextPageUser = user;
+                }
+                if (elements.length / AUCTION_ELEMENT_WIDTH >= pageSize) {
+                    // We are at capacity, return
+                    return (elements, nextPageUser, nextPageUserOffset);
+                }
+            }
+        }
+    }
```

Comment:
If I understand Solidity correctly, this function works but it relies on return statements being optional. Maybe this is obvious to people with more Solidity experience but it confused me.
If the last page has 0 elements then the for loop will not be run which seemed like a bug to me because it would return. But it turns out that with named return variables you don't need to explicitly return them so the function still works correctly.
You don't have to change anything. I just wanted to comment on that.

Reply:
Yes, you are right that this is pretty solidity specific. I'm happy to add a return statement at the end to make it explicit. 

I wanted to reuse the variables to make sure we are not doing more mem copy than necessary (and I'm not sure how the solidity compiler optimizes return values)

Reply:
No need to change. It is intentional and a normal part of solidity.